src/errors/errors.ts

```diff
@@ -1,153 +1,201 @@
-/**
- * Base class for all errors thrown by the sdk
- */
-abstract class SdkError extends Error {
-  protected constructor(message: string) {
-    super(message);
-  }
+export enum MomentoErrorCode {
+  // Invalid argument passed to Momento client
+  INVALID_ARGUMENT_ERROR = 'INVALID_ARGUMENT_ERROR',
+  // Service returned an unknown response
+  UNKNOWN_SERVICE_ERROR = 'UNKNOWN_SERVICE_ERROR',
+  // Cache with specified name already exists
+  ALREADY_EXISTS_ERROR = 'ALREADY_EXISTS_ERROR',
+  // Cache with specified name doesn't exist
+  NOT_FOUND_ERROR = 'NOT_FOUND_ERROR',
+  // An unexpected error occurred while trying to fulfill the request
+  INTERNAL_SERVER_ERROR = 'INTERNAL_SERVER_ERROR',
+  // Insufficient permissions to perform operation
+  PERMISSION_ERROR = 'PERMISSION_ERROR',
+  // Invalid authentication credentials to connect to cache service
+  AUTHENTICATION_ERROR = 'AUTHENTICATION_ERROR',
+  // Request was cancelled by the server
+  CANCELLED_ERROR = 'CANCELLED_ERROR',
+  // Request rate exceeded the limits for the account
+  LIMIT_EXCEEDED_ERROR = 'LIMIT_EXCEEDED_ERROR',
+  // Request was invalid
+  BAD_REQUEST_ERROR = 'BAD_REQUEST_ERROR',
+  // Client's configured timeout was exceeded
+  TIMEOUT_ERROR = 'TIMEOUT_ERROR',
+  // Server was unable to handle the request
+  SERVER_UNAVAILABLE = 'SERVER_UNAVAILABLE',
+  // A client resource (most likely memory) was exhausted
+  CLIENT_RESOURCE_EXHAUSTED = 'CLIENT_RESOURCE_EXHAUSTED',
+  // System is not in a state required for the operation's execution
+  FAILED_PRECONDITION_ERROR = 'FAILED_PRECONDITION_ERROR',
+  // Unknown error has occurred
+  UNKNOWN_ERROR = 'UNKNOWN_ERROR',
 }
 
-abstract class ClientSdkError extends SdkError {
-  protected constructor(message: string) {
-    super(message);
+export class MomentoGrpcErrorDetails {
+  public readonly code: number;
+  public readonly details: string;
+  public readonly metadata: object | null;
+  constructor(code: number, details: string, metadata: object | null) {
+    this.code = code;
+    this.details = details;
+    this.metadata = metadata;
   }
 }
 
+export class MomentoErrorTransportDetails {
+  public readonly grpc: MomentoGrpcErrorDetails;
+  constructor(grpc: MomentoGrpcErrorDetails) {
+    this.grpc = grpc;
+  }
+}
 /**
- * Represents errors thrown when invalid parameters are passed to the Momento Cache
+ * Base class for all errors thrown by the sdk
  */
-export class InvalidArgumentError extends ClientSdkError {
-  constructor(message: string) {
-    super(message);
-    this.name = 'InvalidArgumentError';
+export abstract class SdkError extends Error {
+  public readonly errorCode: MomentoErrorCode;
+  public readonly messageWrapper: string;
+  public readonly transportDetails: MomentoErrorTransportDetails;
+  public readonly stack: string | undefined;
```

Comment:
Why is this undefined instead of null?

Reply:
This Is The Way in typescript :)

![this is the way](https://i.pinimg.com/736x/41/d2/4b/41d24b25710d7746b22a7eb34316952c.jpg)